device.go

```diff
@@ -229,37 +231,174 @@ func virtioBlkDeviceHandler(_ context.Context, device pb.Device, spec *pb.Spec,
+
+	overridePath := filepath.Join(sysPath, "driver_override")
+	err := ioutil.WriteFile(overridePath, []byte("vfio-pci"), 0200)
+	if err != nil {
+		return fmt.Errorf("Failed to write %s: %s", overridePath, err)
+	}
+
+	driverPath := filepath.Join(sysPath, "driver")
+	driver, err := os.Readlink(driverPath)
+	if os.IsNotExist(err) {
+		// This just means the device isn't bound to any driver
+		driver = ""
+	} else if err != nil {
+		return fmt.Errorf("Failed to readlink %s: %s", driverPath, err)
+	}
+
+	fieldLogger.Debugf("VFIO: Device previously bound to '%s'", driver)
+
+	if driver != "vfio-pci" && driver != "" {
+		fieldLogger.Debugf("VFIO: Unbinding from %s", driver)
+		unbindPath := filepath.Join(sysPath, bdf, "driver", "unbind")
+		err = ioutil.WriteFile(unbindPath, []byte(bdf), 0200)
+		if err != nil {
+			return fmt.Errorf("Failed to write %s: %s:", unbindPath, err)
+		}
+	}
+
+	err = ioutil.WriteFile(pciProbeFile, []byte(bdf), 0200)
+	if err != nil {
+		return fmt.Errorf("Failed to write %s: %s", pciProbeFile, err)
+	}
+
+	fieldLogger.Info("VFIO: Rebind complete")
+	return nil
+}
+
+func vfioDeviceHandler(ctx context.Context, device pb.Device, spec *pb.Spec, s *sandbox, devIdx devIndex) error {
+	fieldLogger := agentLog.WithField("host-dev", device.ContainerPath)
+
+	var group string
+
+	for _, opt := range device.Options {
+		l := strings.Split(opt, "=")
+		if len(l) != 2 {
+			return fmt.Errorf("Malformed VFIO option \"%s\"", opt)
+		}
+
+		hostBdf := l[0]
+		guestPath := l[1]
+
+		guestAddr, err := getDevicePCIAddress(guestPath)
+		if err != nil {
+			return err
+		}
+
+		l = strings.Split(guestAddr, "/")
+		guestBdf := l[len(l)-1]
+
+		s.vfioDeviceMap[hostBdf] = guestBdf
+
+		err = vfioRebindDevice(guestBdf)
+		if err != nil {
+			return err
+		}
+
+		groupPath := filepath.Join(sysBusPrefix, guestBdf, "iommu_group")
+		groupLink, err := os.Readlink(groupPath)
+		if err != nil {
+			return fmt.Errorf("Failed to read %s: %s", groupPath, err)
+		}
+
+		if group == "" {
+			group = filepath.Base(groupLink)
+		} else if group != filepath.Base(groupLink) {
+			// If PCI devices associated with the same
+			// VFIO device (and therefore group) in the
+			// host don't end up in the same group in the
+			// guest, something has gone horribly wrong
+			return fmt.Errorf("%s is not in guest IOMMU group %s", guestBdf, group)
+		}
+		fieldLogger.WithField("host-bdf", hostBdf).WithField("guest-bdf", guestBdf).Debug("VFIO: Device complete")
+	}
+
+	vmpath, err := getDeviceName(s, filepath.Join("vfio", group))
+	if err != nil {
+		return fmt.Errorf("Couldn't find device for guest VFIO group %s: %s",
+			group, err)
+	}
+	device.VmPath = vmpath
+
+	fieldLogger.WithField("guest-dev", device.VmPath).Debug("VFIO: Group complete")
+
+	return updateSpecDevice(spec, devIdx, device.ContainerPath, device.VmPath, device.VmPath)
+}
+
+func VfioTranslateEnv(s *sandbox, env []string) []string {
```

Comment:
Ah, this is where the magic happens. Now I finally get it. Smart.

I would add a comment showing the kind of inputs that you expects. Basically, something like "the container receives information about the PCI devices through environment variables that look like `PCIDEVICE_ 0000:00:14.0=...` etc, and explain the replacement you are doing.

Reply:
I believe that function `VfioTranslateEnv` has no unit test.